flang/lib/Optimizer/Transforms/StackArrays.cpp

```diff
@@ -361,6 +361,13 @@ void AllocationAnalysis::visitOperation(mlir::Operation *op,
   } else if (mlir::isa<fir::FreeMemOp>(op)) {
     assert(op->getNumOperands() == 1 && "fir.freemem has one operand");
     mlir::Value operand = op->getOperand(0);
+
+    // Note: StackArrays is scheduled in the pass pipeline after lowering hlfir
+    // to fir. Therefore, we only need to handle `fir::DeclareOp`s.
+    if (auto declareOp =
```

Comment:
I am not sure this a proper use of the DFA framework in this case or not. In particular, should I instead propagate the information through the `fir.declare` op instead of doing this indirect check of the operand?

Reply:
I don't have the context on the `fir` dialect to comment on the memory effects model here, but what I can say is this *should* be fine. What you want is if the allocation state of the result of `declareOp` is modified, that this function gets re-invoked on the right user, which the DFA framework should ensure happens properly.

Reply:
I agree this is fine and a lot simpler than tracking both references to the `allocated` variable